kernel/src/engine/arrow_expression.rs

```diff
@@ -135,6 +135,82 @@ fn column_as_struct<'a>(
+/// associated types as well. This returns an `Ok` of the arrow type that can be used if the types
+/// match, or an error if the types do not match.
+fn ensure_data_types(
+    kernel_type: &DataType,
+    arrow_type: &ArrowDataType,
+) -> DeltaResult<ArrowDataType> {
+    match (kernel_type, arrow_type) {
+        (DataType::Primitive(_), _) if arrow_type.is_primitive() => Ok(arrow_type.clone()),
+        (DataType::Primitive(PrimitiveType::Boolean), ArrowDataType::Boolean) => {
+            // bools aren't primitive in arrow
+            Ok(arrow_type.clone())
+        }
+        (DataType::Primitive(PrimitiveType::String), ArrowDataType::Utf8) => {
+            // strings aren't primitive in arrow
+            Ok(arrow_type.clone())
+        }
+        (DataType::Primitive(PrimitiveType::Binary), ArrowDataType::Binary) => {
+            // binary isn't primitive in arrow
+            Ok(arrow_type.clone())
+        }
+        (
+            DataType::Primitive(PrimitiveType::Decimal(kernel_prec, kernel_scale)),
+            ArrowDataType::Decimal128(arrow_prec, arrow_scale),
+        ) if arrow_prec == kernel_prec && arrow_scale == kernel_scale => {
+            // decimal isn't primitive in arrow
+            Ok(arrow_type.clone())
+        }
+        (DataType::Array(inner_type), ArrowDataType::List(arrow_list_type)) => {
+            let kernel_array_type = &inner_type.element_type;
+            let arrow_list_type = arrow_list_type.data_type();
+            let _ = ensure_data_types(kernel_array_type, arrow_list_type)?;
+            Ok(arrow_type.clone())
+        }
+        (DataType::Map(kernel_map_type), ArrowDataType::Map(arrow_map_type, _)) => {
+            if let ArrowDataType::Struct(fields) = arrow_map_type.data_type() {
+                let mut fiter = fields.iter();
+                if let Some(key_type) = fiter.next() {
+                    let _ = ensure_data_types(&kernel_map_type.key_type, key_type.data_type())?;
+                } else {
+                    return Err(make_arrow_error(
+                        "Arrow map struct didn't have a key type".to_string(),
+                    ));
+                }
+                if let Some(value_type) = fiter.next() {
+                    let _ = ensure_data_types(&kernel_map_type.value_type, value_type.data_type())?;
+                } else {
+                    return Err(make_arrow_error(
+                        "Arrow map struct didn't have a value type".to_string(),
+                    ));
+                }
+                Ok(arrow_type.clone())
+            } else {
+                Err(make_arrow_error(
+                    "Arrow map type wasn't a struct.".to_string(),
+                ))
+            }
+        }
+        (DataType::Struct(kernel_fields), ArrowDataType::Struct(arrow_fields)) => {
+            for (kernel_field, arrow_field) in kernel_fields.fields().zip(arrow_fields.iter()) {
```

Comment:
zip will truncate to the shorter input; a Bad Thing if the two schemas have different field counts